src/backend/columnar/columnar_customscan.c

```diff
@@ -491,15 +515,324 @@ RelationIdGetNumberOfAttributes(Oid relationId)
+			if (var->varno == rel->relid)
+			{
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * FilterPushdownClauses filters for clauses that are candidates for pushing
+ * down into rel.
+ */
+static List *
+FilterPushdownClauses(RelOptInfo *rel, List *inputClauses)
+{
+	List *filteredClauses = NIL;
+	ListCell *lc;
+	foreach(lc, inputClauses)
+	{
+		RestrictInfo *rinfo = (RestrictInfo *) lfirst(lc);
+
+		if (rinfo->pseudoconstant ||
+			!bms_is_member(rel->relid, rinfo->required_relids) ||
+			!CheckPushdownClause(rel, rinfo->clause))
+		{
+			continue;
+		}
+
+		filteredClauses = lappend(filteredClauses, rinfo);
+	}
+
+	return filteredClauses;
+}
+
+
+static bool
+PushdownJoinClauseMatches(PlannerInfo *root, RelOptInfo *rel,
+						  EquivalenceClass *ec, EquivalenceMember *em,
+						  void *arg)
+{
+	return true;
+}
+
+
+/*
+ * FindPushdownJoinClauses finds join clauses, including those implied by ECs,
+ * that may be pushed down.
+ */
+static List *
+FindPushdownJoinClauses(PlannerInfo *root, RelOptInfo *rel)
+{
+	List *joinClauses = copyObject(rel->joininfo);
+
+	/*
+	 * XXX: Here we are generating the clauses just so we can later extract
+	 * the interesting relids. This can probably be made more efficient.
+	 */
+	List *ecClauses = generate_implied_equalities_for_column(
+		root, rel, PushdownJoinClauseMatches, NULL,
+		rel->lateral_referencers);
+	List *allClauses = list_concat(joinClauses, ecClauses);
+
+	return FilterPushdownClauses(rel, allClauses);
+}
+
+
+/*
+ * FindCandidateRelids identifies candidate rels for parameterization from the
+ * list of join clauses.
+ *
+ * Some rels cannot be considered for parameterization, such as a partitioned
+ * parent of the given rel. Other rels are just not useful because they don't
+ * appear in a join clause that could be pushed down.
+ */
+static Relids
+FindCandidateRelids(PlannerInfo *root, RelOptInfo *rel, List *joinClauses)
+{
+	Relids candidateRelids = NULL;
+	ListCell *lc;
+	foreach(lc, joinClauses)
+	{
+		RestrictInfo *rinfo = (RestrictInfo *) lfirst(lc);
+
+		candidateRelids = bms_add_members(candidateRelids,
+										  rinfo->required_relids);
+	}
+
+	candidateRelids = bms_del_member(candidateRelids, rel->relid);
+	return candidateRelids;
+}
+
+
+/*
+ * combinations() calculates the number of combinations of n things taken k at
+ * a time. Intended for small numbers. If it overflows to inf, that's OK (but
+ * be careful with the calculation to not divide by 0 or inf).
+ */
+static double
+combinations(int n, int k)
+{
+	/*
+	 * combinations(n,k) = n!/(k! * (n-k)!)
+	 *
+	 * = (n * (n-1) * ... * (b+1)) / (2 * 3 * ... * a)
+	 *    where a = min(k, n-k) and b = max(k, n-k)
+	 */
+	Assert(n >= k);
+
+	int a = Min(k, n - k);
+	int b = Max(k, n - k);
+
+	double v = 1;
+
+	/* numerator: n * (n-1) ... * (b+1) */
+	for (int i = n; i > b; i--)
+	{
+		v *= i;
+	}
+
+	/* denominator: 2 * 3 * ... * a */
+	for (int i = 2; i <= a; i++)
+	{
+		v /= i;
+	}
```

Comment:
codecov complains that we don't test this loop, probably because `Min(k, n - k)`  becomes `< 2` in all tests.

Reply:
Fixed.